wicket/src/state/status.rs

```diff
@@ -0,0 +1,141 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! State related to downstream service connectivity
+//!
+//! wicketd connectivity is measured directly, and MGS connectivity is proxied
+//! through wicketd.
+
+use crate::ui::defaults::style;
+use std::time::{Duration, Instant};
+use tui::style::Style;
+use tui::text::Span;
+
+/// A status bar shown at the bottom of the screen.
+#[derive(Debug)]
+pub struct ServiceStatus {
+    pub wicketd_liveness: LivenessState,
+    pub mgs_liveness: LivenessState,
+    last_redraw_at: Option<Instant>,
+}
+
+impl ServiceStatus {
+    pub fn new() -> Self {
+        Self {
+            // Wicketd is polled every 500ms by wicket. Setting a 1 second
+            // liveness threshold means that under normal operation it should
+            // never flip to delayed.
+            wicketd_liveness: LivenessState::new(Duration::from_secs(1)),
+            // MGS is polled every 10 seconds by wicketd. Set a 11 second
+            // threshold to account for wicket -> wicketd and wicketd -> MGS
+            // delay.
+            mgs_liveness: LivenessState::new(Duration::from_secs(11)),
+            last_redraw_at: None,
+        }
+    }
+
+    pub fn reset_wicketd(&mut self, elapsed: Duration) {
+        self.wicketd_liveness.reset(elapsed);
+        // Force a redraw.
+        self.last_redraw_at = None;
+    }
+
+    pub fn reset_mgs(&mut self, elapsed: Duration) {
+        self.mgs_liveness.reset(elapsed);
+        // Force a redraw.
+        self.last_redraw_at = None;
+    }
+
+    /// Returns true if a redraw needs to happen, resetting the internal timer.
+    pub fn should_redraw(&mut self) -> bool {
+        if let Some(instant) = &mut self.last_redraw_at {
+            let elapsed = instant.elapsed();
+            if elapsed >= Duration::from_secs(1) {
+                *instant = Instant::now();
+                true
+            } else {
+                false
+            }
+        } else {
+            // Initialize the last-redraw timer.
+            self.last_redraw_at = Some(Instant::now());
+            true
+        }
+    }
+}
+
+/// Tracker used by a single instance of liveness.
+#[derive(Debug)]
+pub struct LivenessState {
+    // Some means that the stopwatch hasn't yet been initialized.
```

Comment:
Should this be "`None` means ..."?

Reply:
I think so. Fixed in [fd4ea2d](https://github.com/oxidecomputer/omicron/pull/2440/commits/fd4ea2d1c6e5f6a298b987e48eeefb5ee6bbeaf0)